### PR TITLE
Fixed #24254 -- Fixed queries using the __in lookup with querysets using distinct() and order_by().

### PR DESCRIPTION
**Pull request by charettes**
https://code.djangoproject.com/ticket/24254

### CODE REVIEW

**Comment by charettes** [comment]
buildbot, test on oracle.

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -518,6 +518,11 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
             if for_update_part and not self.connection.features.for_update_after_from:
                 result.append(for_update_part)
 
+            if self.query.subquery and extra_select:
+                result = [
+                    'SELECT', ', '.join(out_cols[:-len(extra_select)]), 'FROM ('
+                ] + result + [') AS subquery']
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Note to self, I also need to consider `select` params here.

**Comment by charettes** [comment]
This ended up not having to be tested on MySQL and Oracle as they both don't support [sliced subqueries](https://github.com/django/django/search?utf8=%E2%9C%93&q=allow_sliced_subqueries&type=).

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -518,6 +518,24 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                cols_sql = []
+                cols_params = []
+                for _, (s_sql, s_params), alias in self.select:
+                    if alias:
+                        s_sql = self.connection.ops.quote_name(alias)
+                    elif with_col_aliases:
+                        s_sql = 'Col%d' % col_idx
+                        col_idx += 1
+                    else:
+                        cols_params.extend(s_params)
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: I'm trying to find  a case to hit this branch but I can't.

File: django/db/models/sql/compiler.py [hunk 3]
```diff
@@ -518,6 +518,13 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
             if for_update_part and not self.connection.features.for_update_after_from:
                 result.append(for_update_part)
 
+            if self.query.subquery and extra_select:
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: Would it be helpful to comment this `if` to explain when it's used?

File: tests/queries/tests.py [hunk 4]
```diff
@@ -2064,6 +2064,27 @@ def test_sliced_delete(self):
         DumbCategory.objects.filter(id__in=DumbCategory.objects.order_by('-id')[1:]).delete()
         self.assertEqual(set(DumbCategory.objects.values_list('id', flat=True)), {3})
 
+    def test_distinct_ordered_sliced_subquery(self):
+        self.assertSequenceEqual(
+            DumbCategory.objects.filter(
+                id__in=DumbCategory.objects.distinct().order_by('id')[0:2],
+            ).order_by('id').values_list('id', flat=True), [1, 2]
+        )
+        # Explicit value.
```
Comment (timgraham) [inline tests/queries/tests.py]: Does "Explicit value" means "using values('id')"?

File: tests/queries/tests.py [hunk 5]
```diff
@@ -2064,6 +2064,27 @@ def test_sliced_delete(self):
...
+        self.assertSequenceEqual(
+            DumbCategory.objects.filter(
+                id__in=DumbCategory.objects.distinct().order_by('-id').values('id')[0:2],
+            ).order_by('id').values_list('id', flat=True), [3, 4]
+        )
+        # Annotated value.
```
Comment (timgraham) [inline tests/queries/tests.py]: I only see the last assertion failing with the fix removed. Should the other assertions be added in a separate commit?

**Comment by timgraham** [comment]
Perhaps using a less technical commit message like "Fixed queries using the __in lookup with querysets using distinct() and order_by()." (adapted from the ticket's description, you might improve it) could be useful also. The current commit message might be more useful as a code comment.

File: django/db/models/sql/compiler.py (hunk 3 above)
Comment (charettes) [inline django/db/models/sql/compiler.py]: I think it does. I'll add a comment.

File: tests/queries/tests.py (hunk 4 above)
Comment (charettes) [inline tests/queries/tests.py]: Exactly, i'll adjust the comment.

File: tests/queries/tests.py (hunk 5 above)
Comment (charettes) [inline tests/queries/tests.py]: Ahh they use to fail but they are passing because we're using `id` for both ordering and selecting. I'll adjust that.

**Comment by charettes** [comment]
@timgraham adjusted the patch based on your input. Let me know if the comments are too verbose.

File: django/db/models/sql/compiler.py [hunk 6]
```diff
@@ -523,6 +523,28 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # If the query is used as a subquery the extra selects would
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: subquery,

File: django/db/models/sql/compiler.py [hunk 7]
```diff
@@ -523,6 +523,28 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # result in more columns than the left hand side expression is
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: left-hand

File: django/db/models/sql/compiler.py [hunk 8]
```diff
@@ -523,6 +523,28 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # expecting. This can happen when a subquery uses a combination
+                # of order_by() and distinct(); forcing the ordering expressions
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: ; -> ,

File: django/db/models/sql/compiler.py [hunk 9]
```diff
@@ -523,6 +523,28 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # to be selected as well. Wrap the query into another subquery
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: in another

File: django/db/models/sql/compiler.py [hunk 10]
```diff
@@ -523,6 +523,28 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
...
+                # to exclude extraneous selects.
+                subselects = []
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: My instinct would be to use a `sub_` prefix for these names (possibly confirmed by an existing sub_params name in other code).

**Pull request merged**